python/RobotControl/RunningWithSSH.py

```diff
@@ -16,27 +25,112 @@ def run_script_on_robot(script: str) -> str:
+        parts = re.split(r'ERROR\s+-', error, maxsplit=1)
+        return parts[1]
 
-    if "error" in error_message:
-        out = f"{result.split(":")[1].strip()}: {result.split(":")[2].strip()}"
-    elif "exception" in error_message:
-        out = result.split(":")[1].strip()
+    #Start thread to check for runtime errors
+    def run_async_checker():
+        asyncio.run(run_script_finished_error_checker(id))
 
-    if out:
-        robot.controller.stop_program()
-        return out
+    error_checker_thread = threading.Thread(target=run_async_checker)
+    error_checker_thread.start()
 
-    robot.controller.stop_program()
+    return "" 
 
-    robot.ssh.write_script(script)
-    robot.controller.load_program()
-    robot.controller.start_program()
-    return "" 
\ No newline at end of file
+async def run_script_finished_error_checker(id: int):
+    """
+    Check for errors in the script after it has finished running.
+    Sends the error to all web clients if an error is found.
+
+        Args:
+            script (str): The script to check.
+    
+        Returns: 
+            None
+    """
+    recurring_logger.debug(f"Robot mode: {robot.controller.robot_mode}")
+    __wait_for_condition(lambda: "Starting" not in robot.controller.robot_mode)
+    
+    recurring_logger.debug(f"Program state: {robot.controller.program_state}")
+    __wait_for_condition(lambda: "PLAYING" not in robot.controller.program_state)
+    
+    latest_errors = robot.ssh.read_lines_from_log(3)
+    recurring_logger.debug(f"Latest errors:\n{latest_errors}")
+    if "Type error" in latest_errors or "Runtime error" in latest_errors:
+        error_line = latest_errors.split("\n")[0]
+        error_text = re.split(r'ERROR\s+-', error_line, maxsplit=1)[1]
+
+        error = retrieve_line_number_text(error_text)
+        __send_error_message_to_web_clients(id, error)
+        return
+    
+    if "New safety mode: SAFETY_MODE_PROTECTIVE_STOP" in latest_errors:
+        error_text = """
+        PROTECTIVE STOP: Reconsider how the robot is moving.
+        You will not be able to run a program for five seconds.
+        """
+        __send_error_message_to_web_clients(id, error_text)
+
+        # Close safety popup
+        robot.controller.close_safety_popup()
+        robot.controller.unlock_protective_stop()
+        return
+
+def retrieve_line_number_text(text: str) -> str:
+    """
+    Modifies the text to include line number. 
+    Uses 29 as the constant for the line difference.
+
+        Args:
+            text (str): The error message.
+    
+        Returns: 
+            str: The modified error message with line number.
+    """
+    line_difference = 29
+    match = re.search(r'\((\d+:\d+)\)', text)
+    if match:
+        line_number = match.group(1) 
+        number = int(line_number.split(":")[0])
+        text = text.replace(match.group(0), f"(line {number - line_difference})")
+        return text
+    return text
```

Comment:
What does this do? More importantly how is the number 29 affected by the addition of e.g. inspection points?

Reply:
It gets cooked by inspection points.

Reply:
I have commented the method call out, and will find another fix somehow